### ISSUE TITLE
exported custom resource exported fields use default values during _init()

### ISSUE THREAD

**Issue by Bropocalypse** [godotengine/godot#68427]
### Godot version

Godot Engine v4.0.beta4.official.e6751549c

### System information

Kubuntu 20.04

### Issue description

When a node has an exported field for a custom resource, exported fields in that resource refer to their default values when _init() is called in that resource.

![image](https://user-images.githubusercontent.com/58696984/200688196-1a67a236-efcf-4fcf-ab88-652eb10d3e1d.png)

### Steps to reproduce

1)create a custom resource script with a field tagged with `@export`
2)create a node script with an exported field for the custom resource type.
3) assign non-default values to the exported resource's exported field in the inspector.
4)print the values of the fields- they'll be default rather than the inspector-assigned values.

This occurs whether the resource is local to the scene or saved in the file system.

### Minimal reproduction project

[customresourcetest.zip](https://github.com/godotengine/godot/files/9965677/customresourcetest.zip)

**Comment by dalexeev** [comment]
Not a bug. `_init` is called when an object is initialized, right after it has been created. `@export` variables are filled after the object is initialized. Use `_ready` or `NOTIFICATION_SCENE_INSTANTIATED` if you need to access child nodes or exported values before the scene enters the tree.

**Comment by Bropocalypse** [comment]
I see.
I think this might be an issue with the docs, then. It's not clearly explained in the various places that the documentation mentions export. While the document _about_ exports mentions that exported values are "saved with the scene", that doesn't imply this behavior to those who aren't aware of it. To those unaware, exported variables appear at face values to be overrides, or data that exists prior to running the project, so even if this behavior is intended it sure _feels_ like a bug.

**Comment by CurtisHughes** [comment]
Not sure if this is still up for debate/consideration, but I would like to add my +1 for this functionality. Would be nice to support "derived" values in Resources. For example:

```gdscript
@export var vitality: int = 1

var health

func _init():
    health = vitality * 10
```

Maybe there is a better way to handle this, but right now my workaround is to use a getter (which I believe recalculates the value every time it is used):

```gdscript
@export var vitality: int = 1

var health:
    get:
        return vitality * 10
```

**Comment by bkmeneguello** [comment]
I think the problem reported here is really a bug. We need to be able to set the initial value of the Resource's properties.
My case is with a class that extends another class, like: `Earth > Globe > Resource`. Globe is a generic class with `radius`, but Earth needs to set this value to 6400000 and that must be done in Godot's editor. If I set the properties in the `_init` nothing happens and the property value is `null` in the `tscn` file. Since it's not possible to overwrite properties in the child classes, at least it should be possible to set them in the `_init` method.

**Comment by Wyxaldir** [comment]
A notification could be added for resources when they're done loading perhaps? Resources don't seem to receive the NOTIFICATION_POSTINITIALIZE notification, which could be used for the purposes people are wanting above I think.
`@export var vitality : int = 1

var hp

## Example, does not work currently
func _notification(what: int) -> void:
	if what == NOTIFICATION_POSTINITIALIZE:
		hp = vitality * 10`

**Comment by Tegiminis** [comment]
Seconding this. It may not be a bug, but it IS annoying.

My particular use case is, like most people I think, in a health component:

```
extends Resource
class_name HealthComponent

var health = 0
@export var health_max = 100

func _init():
	health = health_max
```

Here is a bare minimum for the custom health resource. I want to initialize my character's current health to their max health at scene start. I create a saved resource where I set ``health_max`` to 50 instead of the default 100. On scene start, it uses the default value defined in the script, rather than the exported value in the saved resource.

I would very much prefer if Resources had a ``_ready()`` function, like Nodes, to do this, particularly if ``_init()`` will continue to ignore inspector-defined values.

**Comment by Calinou** [comment]
Adding a new notification should be discussed in a [proposal](https://github.com/godotengine/godot-proposals), as feature proposals are tracked on the Godot proposals repository.

**Comment by Tegiminis** [comment]
Is there a particular reason ``NOTIFICATION_POSTINITIALIZE`` never gets called in a ``Resource``?

**Comment by loteque** [comment]
If I am understanding this correctly would this proposal close this issue for everyone? https://github.com/godotengine/godot-proposals/issues/5994

**Comment by Dinozor** [comment]
Stumbled on this multiple times while working with Resources in different scenarios.
My solution was to use call_deferred():
```gdscript
func _init() -> void:
	call_deferred("_ready") # To have default image shown. Otherwise magenta


func _ready() -> void:
	print("Network texture initialized. Setting default image ...)
	_set_image()
```
But it feels more as a hack and I have no idea how good it is and what problems it may bring in the future.

**Comment by Turtyo** [comment]
Hello, I'm having a similar problem with exported variable when initializing a dictionary with init
```GDScript
class_name StatModifiers extends Resource

var _MODIFIER_KEYS : Dictionary = GlobalVar.MODIFIER_KEYS

@export var modifier_base_dict: Dictionary = {}

@export var permanent_add: int = 0
@export var permanent_multiply: float = 1
@export var temporary_add: int = 0
@export var temporary_multiply: float = 1

func _init( _permanent_add: int = permanent_add, 
	    _permanent_multiply: float = permanent_multiply, 
	    _temporary_add: int = temporary_add, 
	    _temporary_multiply: float = temporary_multiply) -> void:
	modifier_base_dict = {   
					_MODIFIER_KEYS.PERMANENT_ADD:        _permanent_add,  # int
					_MODIFIER_KEYS.PERMANENT_MULTIPLY:   _permanent_multiply,  # float
					_MODIFIER_KEYS.TEMPORARY_ADD:        _temporary_add,  # int
					_MODIFIER_KEYS.TEMPORARY_MULTIPLY:   _temporary_multiply   # float
				}
```
Problem being when I use this from the inspector, I must create the instance of `StatModifiers` to be able to change the `@export` value, but by creating the `StatModifiers` it just initializes it with the default value, and changing the export variables doesn't change the dictionary anymore
Which honestly feels counter-intuitive as others have pointed before.

**Comment by TrieBr** [comment]
Isn't this what setters are for? It seems odd that the properties would be set before the constructor/_init is called. At least in C# the object needs to be constructed before the public properties are set so it makes sense why the behavior is this way.

In all of the examples in this thread, a setter function with an optional backing field would work fine to ensure that derived variables are set properly.

i.e. @Dinozor example:

```
class_name NetworkTexture
extends ImageTexture

var _default;
@export var default: Texture2D  :
    get:
        return _default
    set(value):
        _default = value
        set_image(value.get_image())
```

In my opinion, it's not really clear when `_ready` would be called. Would it be called again if properties change during runtime? That could be inefficient. Setters seem like the most straightforward way to guarantee that derived data is re-computed whenever that property changes.

**Comment by Tegiminis** [comment]
Why would `_ready` be called again if properties change during runtime? It is used all over the place in Godot already and has a clearly defined place in the load hierarchy. It is called when a node "enters the tree"; in other words, as it is loaded into the scene. It occurs after `_init`, which happens when the node (or resource) is initialized into memory. What's confusing is that `@export` variables are only applied AFTER `_init`, which may be intentional but which leaves a gap in the initialization of Resources that should not have to be filled with getter/setter boilerplate or workarounds.

Also, in my particular use-case, I'm not looking for the derived data to be recomputed every time the property changes. I just want a default value that I can set in the inspector and which is applied to the resource when it is loaded (health that is set to max on scene load). You can do this already directly on nodes, but not Resources. Resources do not call `_ready` and do not trigger `NOTIFICATION_POSTINITIALIZE`. It's very unintuitive and contrary to the best practice Godot teaches you for Node scripts, and in my opinion it's a bug.

My personal workaround is I created my own `_ready` within the resource to initialize the defaults:
```
func _ready():
	health = health_max
	armor = armor_max
```

Which I then call within the Node's `_ready`:
```
func _ready():
	animator.play("idle")
	health._ready()
```

**Comment by AThousandShips** [comment]
I'm not sure how that relates to this issue/discussion? Can you elaborate?

**Comment by PinkSerenity** [comment]
> I'm not sure how that relates to this issue/discussion? Can you elaborate?

Well, most of the use cases people describe here are about assigning the value of an exported variable to another as soon as possible. Though in some cases it is modified.

Maybe this would be better?
```GDScript
@export field_name: String: # Field Name will become the name of the field in the editor, but not a variable
	var max_health = field_name
	var health = field_name / 2
	inherited_variable # at least until the support of interfaces
```
Though that basically makes `field_name` behave like a variable again. Then you could easily just make a setter.

**Comment by Tegiminis** [comment]
No. I **specifically** want custom Resources to also utilize `_ready`, for parity with Nodes. This isn't about solving a specific use case (of which there are like 4 or 5 workarounds), it's about fixing Resources.

**Comment by PinkSerenity** [comment]
> No. I _**specifically**_ want custom Resources to also utilize `_ready`, for parity with Nodes. This isn't about solving a specific use case (of which there are like 4 or 5 easy workarounds), it's about fixing Resources. So your suggestion is not helpful.

I'm sorry that it isn't helpful and that I bothered you there. That wasn't my intention. As somebody else mentioned, there is a [proposal](https://github.com/godotengine/godot-proposals/issues/5994) to basically add a method that gets called when the resource and all variables have been initialized. As long as it isn't implemented, I fear that workarounds are the only solution we have. That's also why I suggested another one, because I until we get that method, finding the best workaround for each use case is the only thing we can do. And that I tried to help with, but apparently failed, so I'm again sorry for that.

**Comment by abrossimow** [comment]
Giving this a gentle nudge: I share the same discomfort.

`_init()` not getting inspector set `@export `variables is confusing and unexpected. Especially, when on solid work-around is using `getters `for certain designs is creating expected behavior.
